src/update_agent/mod.rs

```diff
@@ -238,6 +248,53 @@ impl UpdateAgent {
     }
 }
 
+/// Attempt to broadcast msg to all sessions registered in systemd's login manager.
+/// Returns count of sessions broadcasted to.
+fn broadcast(msg: &str) -> Fallible<u8> {
+    // Get list of sessions with users logged in with `loginctl`.
+    let cmdrun = std::process::Command::new("loginctl")
+        .arg("list-sessions")
+        .arg("--output=json")
+        .output()
+        .with_context(|_| "failed to run 'loginctl' binary")?;
+
+    if !cmdrun.status.success() {
+        bail!(
+            "loginctl failed to list current sessions:\n{}",
+            String::from_utf8_lossy(&cmdrun.stderr)
+        );
+    }
+
+    let sessions: Vec<SessionsJSON> = serde_json::from_slice(&cmdrun.stdout)?;
+    let mut sessions_count: u8 = 0;
+
+    // Iterate over sessions and attempt to write to each session's tty.
+    for session in sessions.into_iter() {
+        let user = session.user;
+        let mut tty = session.tty;
+        tty.insert_str(0, "/dev/");
+
+        log::trace!(
+            "Attempting to broadcast a message to user {} at {}",
+            user,
+            tty
+        );
+
+        {
+            let broadcast = format!(
+                "Broadcast message from Zincati at {}:\n{}\n",
+                chrono::Utc::now().format("%a %Y-%m-%d %H:%M:%S %Z"),
+                msg
+            );
+            fs::write(&tty, broadcast)?;
```

Comment:
I'd rather re-arrange this so that in case of errors we do a `log::error!()` but we keep trying the other sessions.

At the end you can return a `Some(tty_written, tty_total)` or a `None` if there were no sessions.

Reply:
I opted to return a Result still, since this would let us propagate any errors and also 'no sessions' doesn't seem like a special enough case to be `None` IMO. What do you think?

Reply:
That's fine as well.